File: solutions/free-pascal/protein-translation/1/ProteinTranslation.pas
unit ProteinTranslation;

{$mode ObjFPC}{$H+}

interface

uses SysUtils;

type
  TStrArray = Array Of String;

function proteins(const strand : string) : TStrArray;

implementation

function peptid(const triple : string) : string;
begin
  case triple of
    'AUG' : result := 'Methionine';
    'UUU', 'UUC' : result := 'Phenylalanine';
    'UUA', 'UUG' : result := 'Leucine';
    'UCU', 'UCC', 'UCA', 'UCG' : result := 'Serine';
    'UAU', 'UAC' : result :='Tyrosine';
    'UGU', 'UGC' : result :='Cysteine';
    'UGG' : result :='Tryptophan';
    'UAA', 'UAG', 'UGA' : result :='STOP';
   else
     result := 'INVALID';
    end;
end;

function proteins(const strand : string) : TStrArray;
var 
  i : integer;
  codon : string;
  amino_acid : string;
begin
  if strand='' then 
    exit([]);
  (* if (length(strand) mod 3) <> 0 then
    raise Exception.Create('Invalid codon'); *)
  i := 1;
  codon := '';
  repeat
    codon := strand[i] + strand[i+1] + strand[i+2];
    if (i+2<length(strand)) and (i+5>length(strand)) then
      raise Exception.Create('Invalid codon');
    
    amino_acid := peptid(codon);    
    if amino_acid = 'STOP' then
      break
    else if amino_acid = 'INVALID' then
      raise Exception.Create('Invalid codon');    
    Insert(amino_acid, result, length(result));
    
    i:= i + 3;
  until (i>=length(strand));
  
  

end;

end.

File: solutions/cpp/protein-translation/2/protein_translation.cpp
#include "protein_translation.h"

#include <vector>
#include <string>
#include <map>
#include <stdexcept>

using std::string;

namespace protein_translation {

    std::map<string,string> amino_map = {
        {"AUG", "Methionine"},
        {"UUU", "Phenylalanine"},
        {"UUC", "Phenylalanine"},
        {"UUA", "Leucine"},
        {"UUG", "Leucine"},
        {"UCU", "Serine"},
        {"UCC", "Serine"},
        {"UCA", "Serine"},
        {"UCG", "Serine"},
        {"UAU", "Tyrosine"},
        {"UAC", "Tyrosine"},
        {"UGU", "Cysteine"},
        {"UGC", "Cysteine"},
        {"UGG", "Tryptophan"},
        {"UAA", "STOP"},
        {"UAG", "STOP"},
        {"UGA", "STOP"}    
    };
    

    std::vector<string> proteins(string rna) {
            std::vector<string> polypeptid;
            for(int i=0; i<=static_cast<int>(rna.length()-3); i+=3){
                string codon = rna.substr(i,3);
                auto search = amino_map.find(codon);
                if(search!=amino_map.end()){
                    if(search->second == "STOP")
                        break;
                    else
                        polypeptid.push_back(search->second);
                } 
                else 
                    throw std::invalid_argument("Invalid codon "
                        + codon);
            }
        return polypeptid;
    }

}  // namespace protein_translation
